Framework/Algorithms/src/MagFormFactorCorrection.cpp

```diff
@@ -0,0 +1,123 @@
+                  "Output workspace name.");
+  std::vector<std::string> keys = getMagneticIonList();
+  declareProperty("IonName", "Cu2",
+                  boost::make_shared<StringListValidator>(keys),
+                  "The name of the ion: an element symbol with a number "
+                  "indicating the valence, e.g. Fe2 for Fe2+ / Fe(II)");
+  declareProperty(make_unique<WorkspaceProperty<>>("FormFactorWorkspace", "",
+                                                   Direction::Output,
+                                                   PropertyMode::Optional),
+                  "If specified the algorithm will create a 1D workspace with "
+                  "the form factor vs Q with a name given by this field.");
+}
+
+void MagFormFactorCorrection::exec() {
+  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  const std::string ionNameStr = getProperty("IonName");
+  const std::string ffwsStr = getProperty("FormFactorWorkspace");
+  const bool isHist = inputWS->isHistogramData();
+  const int64_t numHists = inputWS->getNumberHistograms();
+  const int64_t specSize = inputWS->blocksize();
+
+  // Checks that there is a |Q| axis.
+  int64_t numAxes = inputWS->axes();
+  bool hasQ = false;
+  std::vector<double> Qvals;
+  int64_t iax;
+  for (iax = 0; iax < numAxes; iax++) {
+    Axis *QAxis = inputWS->getAxis(iax);
+    std::string unitID = QAxis->unit()->unitID();
+    if (unitID == "MomentumTransfer") {
+      hasQ = true;
+      // Gets the list of Q values
+      if (isHist || iax > 0) {
+        int64_t nQ = QAxis->length() - 1;
+        for (int64_t iQ = 0; iQ < nQ; iQ++) {
+          Qvals.push_back(0.5 * (QAxis->getValue(static_cast<size_t>(iQ)) +
+                                 QAxis->getValue(static_cast<size_t>(iQ + 1))));
+        }
+      } else {
+        int64_t nQ = QAxis->length();
+        for (int64_t iQ = 0; iQ < nQ; iQ++) {
+          Qvals.push_back(QAxis->getValue(static_cast<size_t>(iQ)));
+        }
+      }
+      break;
+    }
+  }
+  if (!hasQ) {
+    throw std::runtime_error("Expected a MatrixWorkspace with a "
+                             "MomentumTransfer axis. Cannot apply form factor "
+                             "without knowing Q.");
+  }
+
+  // Parses the ion name and get handle to MagneticIon object
+  const MagneticIon ion = getMagneticIon(ionNameStr);
+  // Gets the vector of form factor values
+  std::vector<double> FF;
+  FF.reserve(Qvals.size());
+  for (int64_t iQ = 0; iQ < (int64_t)Qvals.size(); iQ++) {
+    FF.push_back(ion.analyticalFormFactor(Qvals[iQ] * Qvals[iQ], 0, 0));
+  }
+  if (!ffwsStr.empty()) {
+    MatrixWorkspace_sptr ffws = API::WorkspaceFactory::Instance().create(
+        "Workspace2D", 1, Qvals.size(), FF.size());
+    ffws->mutableX(0).assign(Qvals.begin(), Qvals.end());
+    ffws->mutableY(0).assign(FF.begin(), FF.end());
+    API::AnalysisDataService::Instance().addOrReplace(ffwsStr, ffws);
```

Comment:
`AnalysisDataService` shouldn't be used directly. Set the optional property instead:

```
setProperty("FormFactorWorkspace", ffws);
```